inst/include/mode/r/mode.hpp

```diff
@@ -54,21 +54,95 @@ cpp11::sexp mode_solve(SEXP ptr, double end_time) {
 }
 
-double validate_time(cpp11::sexp r_time) {
+template <typename T>
+cpp11::sexp mode_state(SEXP ptr) {
+  T *obj = cpp11::as_cpp<cpp11::external_pointer<T>>(ptr).get();
+  std::vector<double> dat(obj->n_state() * obj->n_particles());
+  obj->state(dat);
+  return state_array(dat, obj->n_state(), obj->n_particles());
+}
+
+std::vector<double> validate_time(cpp11::sexp r_time) {
   cpp11::doubles time = cpp11::as_cpp<cpp11::doubles>(r_time);
   if (time.size() != 1) {
-    cpp11::stop("expected 'time' to be a scalar value");
+    cpp11::stop("Expected 'time' to be a scalar value");
+  }
+  std::vector<double> ret(1);
+  std::copy_n(REAL(time.data()), 1, ret.begin());
+  return ret;
+}
+
+cpp11::integers object_dimensions(cpp11::sexp obj, size_t obj_size) {
+  cpp11::integers dim;
+  auto r_dim = obj.attr("dim");
+  if (r_dim == R_NilValue) {
+    dim = cpp11::writable::integers{static_cast<int>(obj_size)};
+  } else {
+    dim = cpp11::as_cpp<cpp11::integers>(r_dim);
+  }
+  return dim;
+}
+
+std::vector<double> validate_state(cpp11::sexp r_state,
+                              int n_state,
+                              int n_particles) {
+  cpp11::doubles r_state_data = cpp11::as_cpp<cpp11::doubles>(r_state);
+  const size_t state_len = r_state_data.size();
+  auto dim = object_dimensions(r_state, n_state);
+  if (dim.size() > 2) {
+    cpp11::stop("Expected 'state' to have at most 2 dimensions");
+  }
+  if (dim.size() == 2) {
+    if (dim[0] != n_state || dim[1] != n_particles) {
+      cpp11::stop("Expected 'state' to be a %d by %d matrix but was %d by %d",
+                  n_state, n_particles, dim[0], dim[1]);
+    }
   }
-  return time[0];
+  if (dim.size() == 1 && static_cast<int>(state_len) != n_state) {
+    cpp11::stop("Expected 'state' to be a vector of length %d but was length %d",
+                n_state, state_len);
+  }
+  std::vector<double> ret(state_len);
+  std::copy_n(REAL(r_state_data.data()), state_len, ret.begin());
+  return ret;
 }
 
 template <typename T>
-void mode_update_state(SEXP ptr, SEXP r_time) {
+void mode_update_state(SEXP ptr, SEXP r_state, SEXP r_time,
+                       SEXP r_set_initial_state,
+                       SEXP r_reset_step_size) {
   T *obj = cpp11::as_cpp<cpp11::external_pointer<T>>(ptr).get();
+
+  bool set_initial_state = false;
+  if (r_set_initial_state == R_NilValue) {
+    set_initial_state = r_state == R_NilValue &&
+                        r_time != R_NilValue;
+  } else {
+    set_initial_state = cpp11::as_cpp<bool>(r_set_initial_state);
+    if (set_initial_state && r_state != R_NilValue) {
+      cpp11::stop("'set_initial_state' cannot be TRUE unless 'state' is NULL");
+    }
+  }
+
+  bool reset_step_size = false;
+  if (r_reset_step_size == R_NilValue) {
+    reset_step_size = r_time != R_NilValue;
+  } else {
+    reset_step_size = cpp11::as_cpp<bool>(r_reset_step_size);
+  }
+
+  std::vector<double> time;
```

Comment:
If you did:

```
auto time = validate_time(r_time);
```

and rewrote validate_time to do:

```
std::vector<double> validate_time(cpp11::sexp r_time) {
  if (r_time == R_NilValue) {
    return std::vector<double>(0);
  }
  cpp11::doubles time = cpp11::as_cpp<cpp11::doubles>(r_time);
  if (time.size() != 1) {
    cpp11::stop("Expected 'time' to be a scalar value");
  }
  std::vector<double> ret(1);
  std::copy_n(REAL(time.data()), 1, ret.begin());
  return ret;
}
```

and followed that pattern for all 4 inputs this function would clean up quite nicely I think.


```
template <typename T>
void mode_update_state(SEXP ptr, SEXP r_state, SEXP r_time,
                       SEXP r_set_initial_state,
                       SEXP r_reset_step_size) {
  T *obj = cpp11::as_cpp<cpp11::external_pointer<T>>(ptr).get();

  auto set_initial_state = validate_set_initial_state(r_set_initial_state, r_time, r_state);
  auto reset_step_size = validate_reset_step_size(r_reset_step_size, r_time);
  auto time = validate_time(r_time);
  auto state = validate_state(r_state);
  obj->update_state(time, state, set_initial_state, reset_step_size);
}
```